gateway/gateway.go

```diff
@@ -185,18 +200,15 @@ func (g *Gateway) Close() error {
 // Reconnect tries to reconnect forever. It will resume the connection if
 // possible. If an Invalid Session is received, it will start a fresh one.
 func (g *Gateway) Reconnect() {
-	for {
-		if err := g.ReconnectCtx(context.Background()); err != nil {
-			g.ErrorLog(err)
-		} else {
-			return
-		}
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), g.WSTimeout)
+	defer cancel()
+
+	g.ReconnectCtx(ctx)
 }
 
 // ReconnectCtx attempts to reconnect until context expires. If context cannot
 // expire, then the gateway will try to reconnect forever.
-func (g *Gateway) ReconnectCtx(ctx context.Context) (err error) {
+func (g *Gateway) ReconnectCtx(ctx context.Context) {
```

Comment:
This section is missing a check for `ReconnectTimeout`. If the timeout is 0, then reconnection should continue inside the loop.

The code should be something like this:

```go
if g.ReconnectTimeout > 0 {
    tctx, cancel := context.WithTimeout(ctx, g.ReconnectTimeout)
    defer cancel()

    ctx = tctx
}
```